teuthology/orchestra/remote.py

```diff
@@ -518,49 +520,59 @@ def move_file(self, src, dst, sudo=False, mode=None, owner=None,
             args += ' && ' + chown + ' ' + owner + ' ' + dst
         self.run(args=args)
 
+    @contextmanager
+    def _maybe_copy_from(self, sudo, path):
+        try:
+            if sudo:
+                temp_path = self.mktemp()
+                self.sh(f'sudo cp ${path} ${temp_path}')
+                self.sh(f'sudo chmod 0666 ${temp_path}')
+                yield temp_path
+            else:
+                yield path
+        finally:
+            if sudo:
+                self.sh(f'rm ${temp_path}')
+
     def read_file(self, path, sudo=False, stdout=None,
-                              offset=0, length=0):
+                              offset=0, length=-1):
         """
         Read data from remote file
 
         :param path:    file path on remote host
         :param sudo:    use sudo to read the file, defaults False
         :param stdout:  output object, defaults to io.BytesIO()
         :param offset:  number of bytes to skip from the file
-        :param length:  number of bytes to read from the file
-
-        :raises: :class:`FileNotFoundError`: there is no such file by the path
-        :raises: :class:`RuntimeError`:      unexpected error occurred
-
+        :param length:  number of bytes to read from the file, if the length is
+                        negative or omitted, read all data until EOF is reached
         :returns: the file contents in bytes, if stdout is `io.BytesIO`, by default
         :returns: the file contents in str, if stdout is `io.StringIO`
         """
-        dd = 'sudo dd' if sudo else 'dd'
-        args = dd + ' if=' + path + ' of=/dev/stdout'
-        iflags=[]
-        # we have to set defaults here instead of the method's signature,
-        # because python is reusing the object from call to call
-        stdout = stdout or BytesIO()
-        if offset:
-            args += ' skip=' + str(offset)
-            iflags += 'skip_bytes'
-        if length:
-            args += ' count=' + str(length)
-            iflags += 'count_bytes'
-        if iflags:
-            args += ' iflag=' + ','.join(iflags)
-        args = 'set -ex' + '\n' + args
-        proc = self.run(args=args, stdout=stdout, stderr=StringIO(), check_status=False)
-        if proc.returncode:
-            if 'No such file or directory' in proc.stderr.getvalue():
-                raise FileNotFoundError(errno.ENOENT,
-                        f"Cannot find file on the remote '{self.name}'", path)
+        log.debug(f'read_file({path}, {offset}, {length}')
+
+        with self._maybe_copy_from(sudo, path) as src_path:
+            sftp = self.ssh.open_sftp()
+            with sftp.open(src_path, 'r') as f:
```

Comment:
Why do we switch to sftp? We already have sftp copy methods.

Reply:
it's simpler to use SFTP. actually the same question also applies to "dd".

Reply:
Easy, we use dd here instead of cat because it allow more file operations. Please do not switch to sftp here.

Reply:
if you want to fix this issue, please go ahead.